File: problems-extend-adapt/3101-3200/3121_letters-in-both-cases-ii/solution.go
func countDualCaseLettersII(word string) int {
	// Special means every lowercase occurrence sits before the first
	// uppercase one, i.e. last-lower index < first-upper index; both
	// positions per letter are captured in a single pass.
	firstUpper := make([]int, 26)
	lastLower := make([]int, 26)
	for k := range firstUpper {
		firstUpper[k] = -1
		lastLower[k] = -1
	}
	for position := 0; position < len(word); position++ {
		ch := word[position]
		if ch >= 'a' {
			lastLower[ch-'a'] = position
		} else if firstUpper[ch-'A'] == -1 {
			firstUpper[ch-'A'] = position
		}
	}
	count := 0
	for k := 0; k < 26; k++ {
		if firstUpper[k] != -1 && lastLower[k] != -1 && lastLower[k] < firstUpper[k] {
			count++
		}
	}
	return count
}

File: problems-extend-adapt/3101-3200/3121_letters-in-both-cases-ii/solution.py
class Solution:
    def countDualCaseLettersII(self, word: str) -> int:
        # Special means every lowercase occurrence sits before the first
        # uppercase one, i.e. last-lower index < first-upper index; both
        # positions per letter are captured in a single pass.
        first_upper = [-1] * 26
        last_lower = [-1] * 26
        for position, ch in enumerate(word):
            if ch.islower():
                last_lower[ord(ch) - ord("a")] = position
            elif first_upper[ord(ch) - ord("A")] == -1:
                first_upper[ord(ch) - ord("A")] = position
        return sum(first_upper[k] != -1 and last_lower[k] != -1 and last_lower[k] < first_upper[k] for k in range(26))
